src/IO/HTTPCommon.cpp

```diff
@@ -238,13 +257,7 @@ namespace
 
                     if (resolve_host)
                     {
-                        /// Host can change IP
-                        const auto ip = DNSResolver::instance().resolveHost(host).toString();
-                        if (ip != session->getHost())
-                        {
-                            session->reset();
-                            session->setHost(ip);
-                        }
```

Comment:
all of this looks like a mess to me...
There is two variables in play - if host is a name or address, and if resolve_host set or not.
The problem apparently can only exist when host is a name and resolve_host is set:
- previously https_session->_resolved_host is set to resolved IP
- https_session->_host is set to host name
- original result: IP is checked against host name - not equal - https_session->_host is changed, but doing nothing, because https_session->_resolved_host is set and used on connect
- your fix result: IP checked against https_session->_resolved_host, and may work, however Poco has it's own DNS cache in DNSResolver, and if it's enabled then this solution will fail also.

And in test you are checking with IP addresses - not host name with changed IP - I don't think it tests the issue...

Reply:
We must not change `_host` for `HTTPSClientSession` because it is used for SNI, as desribed earlier. The old code was correct for HTTP session but not for HTTPS for which `_resolved_host`  [was introduced](https://github.com/ClickHouse/ClickHouse/commit/77fb1b84acb88f6b211a44c605773443758e5f18). We should always update `_resolved_host` if it was set.


> And in test you are checking with IP addresses - not host name with changed IP - I don't think it tests the issue...

Could you elaborate this ? 

[These](https://github.com/ClickHouse/ClickHouse/pull/50240/files#diff-eaa6b0991a642bebf69b10d073c33171d41db5cd9155dd20612bc2b0599c0572R83-R84:~:text=%23%20We%20change%20source%20node%20ip) statements restart replica with old host name but new IP address.

Reply:
ok, seems legit
but probably you will need also to manipulate with DNS cache settings to solve this problem